libsnark/zk_proof_systems/plonk/prover.tcc

```diff
@@ -248,14 +249,71 @@ round_three_out_t<ppT> plonk_prover<ppT, transcript_hasher>::round_three(
     polynomial<Field> cqO;
     libfqfft::_polynomial_multiplication<Field>(
         cqO, round_one_out.W_polys_blinded[c], srs.Q_polys[O]);
+
+    std::shared_ptr<libfqfft::evaluation_domain<Field>> domain =
+        libfqfft::get_evaluation_domain<Field>(srs.num_gates);
+
+    // Compute PI polynomial from the PI_wire_index (stored in the srs) and
+    // the PI_value (stored in the witness). By convention of this
+    // implementation the PI values are stored in the right input wire w_R (and
+    // not in the left input wire w_L as in [GWC19]). Recall that the witness w
+    // is composed of left input w_L, right input w_R and output wires w_O (in
+    // this order) and so is the concatenation of w_L || w_R || w_O = w. Each
+    // vector w_L, w_R and w_O of wire values is of length srs.num_gates and so
+    // in order to get the value of the i-th PI[i] located at PI_wire_index[i]
+    // of w we need to do a modulo srs.num_gates operation (to skip the first
+    // srs.num_gates values corresponding to w_L). This is the reason for the
+    // following modulo srs.num_gates operation.
+    //
+    // EXAMPLE
+    //
+    // Suppose that we have a circuit of 8 gates and the witness is w = w_L ||
+    // w_R || w_O where (leftmost position corresponds to index 0):
+    //
+    // w_L = [ 3, 9,27, 1, 1,30, 0, 0]
+    // w_R = [ 3, 3, 3, 5,35, 5, 0, 0]
+    // w_O = [ 9,27,30, 5,35,35, 0, 0]
+    //
+    // so (leftmost position corresponds to index 0):
+    //
+    // w = [ 3,9,27,1,1,30,0,0,3,3,3,5,35,5,0,0,9,27,30,5,35,35,0,0]
+    //
+    // In this example circuit we have a single public input (PI) 35 first
+    // appearing at position 12 in w (counting from 0). In this case the vector
+    // srs.PI_wire_index will contain a single element srs.PI_wire_index[0]
+    // = 12.
+    //
+    // The value of the PI is extracted from the witness as PI_value =
+    // w[srs.PI_wire_index[0]] = w[12] = 35
+    //
+    // To obtain the index of the PI in w_R we do a modulo num_gates (= 8)
+    // operation to skip the w_L vector (first 8 entries in w). Note that this
+    // index also corresponds to the power of x in the PI polynomial:
+    //
+    // power_of_x = srs.PI_wire_index[0] % num_gates = 12 % 8 = 4
+    //
+    // Finally the PI polynomial is computed as
+    //
+    // PI_poly(x) = PI_value x^power_of_x = 35 x^4
```

Comment:
PI should be the iFFT of this (i.e. the iFFT of the public inputs at their `indices % srs.num_gates`).  The code seems correct, but this comment doesn't quite correspond to it.  The logic that is specific to this PI polynomial is done in this function, and then `plonk_compute_public_input_polynomial` is called, which is actually just the very generic part of the computation.

Reply:
That is correct. Fixed.